doc/man/mvcd.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.8.
.TH MVCD "1" "November 2021" "mvcd v0.1.4.0" "User Commands"
.SH NAME
mvcd \- manual page for mvcd v0.1.4.0
.SH DESCRIPTION
MicroVisionChain version v0.1.4.0\-dev
.SS "Usage:"
.TP
mvcd [options]
Start MicroVisionChain Daemon
.SH OPTIONS
.HP
\-?
.IP
Print this help message and exit
.HP
\fB\-version\fR
.IP
Print version and exit
.HP
\fB\-alertnotify=\fR<cmd>
.IP
Execute command when a relevant alert is received or we see a really
long fork (%s in cmd is replaced by message)
.HP
\fB\-blocknotify=\fR<cmd>
.IP
Execute command when the best block changes (%s in cmd is replaced by
block hash)
.HP
\fB\-assumevalid=\fR<hex>
.IP
If this block is in the chain assume that it and its ancestors are valid
and potentially skip their script verification (0 to verify all,
default:
000000000000000000e45ad2fbcc5ff3e85f0868dd8f00ad4e92dffabe28f8d2,
testnet:
0000000000327972b8470c11755adf8f4319796bafae01f5a6650490b98a17db)
.HP
\fB\-conf=\fR<file>
.IP
Specify configuration file (default: mvc.conf)
.HP
\fB\-daemon\fR
.IP
Run in the background as a daemon and accept commands
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.HP
\fB\-dbcache=\fR<n>
.IP
Set database cache size in megabytes (4 to 16384, default: 450). The
value may be given in megabytes or with unit (B, KiB, MiB, GiB).
.IP
Set cache size for database holding a list of frozen transaction outputs
in bytes (default: 128000)
.HP
\fB\-genesisactivationheight\fR
.IP
Set block height at which genesis should be activated. (default:
620538).
.HP
\fB\-loadblock=\fR<file>
.IP
Imports blocks from external blk000??.dat file on startup
.HP
\fB\-maxmempool=\fR<n>
.IP
Keep the resident size of the transaction memory pool below <n>
megabytes (default: 1000,  must be at least 300). The value may
be given in megabytes or with unit (B, kB, MB, GB).
.HP
\fB\-mempoolmaxpercentcpfp=\fR<n>
.IP
Percentage of total mempool size (ram+disk) to allow for low paying
transactions (0..100) (default: 10)
.HP
\fB\-mempoolexpiry=\fR<n>
.IP
Do not keep transactions in the mempool longer than <n> hours (default:
336)
.HP
\fB\-maxmempoolnonfinal=\fR<n>
.IP
Keep the non\-final transaction memory pool below <n> megabytes (default:
50). The value may be given in megabytes or with unit (B, KiB,
MiB, GiB).
.HP
\fB\-mempoolexpirynonfinal=\fR<n>
.IP
Do not keep transactions in the non\-final mempool longer than <n> hours
(default: 672)
.HP
\fB\-persistmempool\fR
.IP
Whether to save the mempool on shutdown and load on restart (default: 1)
.HP
\fB\-threadsperblock=\fR<n>
.IP
Set the number of script verification threads used when validating
single block (0 to 64, 0 = auto, default: 0)
.HP
\fB\-scriptvalidatormaxbatchsize=\fR<n>
.IP
Set size of script verification batch per thread (1 to 255, default:
128)
.HP
\fB\-maxparallelblocks=\fR<n>
.IP
Set the number of block that can be validated in parallel across all
nodes. If additional block arrive, validation of an old block is
terminated. (1 to 100, default: 4)
.HP
\fB\-maxparallelblocksperpeer=\fR<n>
.IP
Set the number of blocks that can be validated in parallel from a single
peer. If peers sends another block, the validation of it is
delayed. (1 to maxparallelblocks, default: 3)
.HP
\fB\-pid=\fR<file>
.IP
Specify pid file (default: mvcd.pid)
.HP
\fB\-preload=\fR<n>
.IP
If n is set to 1, blockchain state will be preloaded into memory. If n
is 0, no preload will happen. Other values for n are not allowed.
The default value is 0. This option is not supported on Windows
operating systems.
.HP
\fB\-prune=\fR<n>
.IP
Reduce storage requirements by enabling pruning (deleting) of old
blocks. This allows the pruneblockchain RPC to be called to
delete specific blocks, and enables automatic pruning of old
blocks if a target size in MiB is provided. This mode is
incompatible with \fB\-txindex\fR and \fB\-rescan\fR. Warning: Reverting this
setting requires re\-downloading the entire blockchain. (default:
0 = disable pruning blocks, 1 = allow manual pruning via RPC,
>550 = automatically prune block files to stay under the
specified target size in MiB)
.HP
\fB\-reindex\-chainstate\fR
.IP
Rebuild chain state from the currently indexed blocks
.HP
\fB\-reindex\fR
.IP
Rebuild chain state and block index from the blk*.dat files on disk
.HP
\fB\-rejectmempoolrequest\fR
.IP
Reject every mempool request from non\-whitelisted peers (default: 1).
.HP
\fB\-sysperms\fR
.IP
Create new files with system default permissions, instead of umask 077
(only effective with disabled wallet functionality)
.HP
\fB\-txindex\fR
.IP
Maintain a full transaction index, used by the getrawtransaction rpc
call (default: 0)
.HP
\fB\-maxmerkletreediskspace\fR
.IP
Maximum disk size in bytes that can be taken by stored merkle trees.
This size should not be less than default size (default:
184549376 bytes). The value may be given in bytes or with unit
(B, kiB, MiB, GiB).
.HP
\fB\-preferredmerkletreefilesize\fR
.IP
Preferred size of a single datafile containing merkle trees. When size
is reached, new datafile is created. If preferred size is less
than size of a single merkle tree, it will still be stored,
meaning datafile size can be larger than preferred size.
(default: 33554432 bytes). The value may be given in bytes or
with unit (B, kiB, MiB, GiB).
.HP
\fB\-maxmerkletreememcachesize\fR
.IP
Maximum merkle trees memory cache size in bytes. For faster responses,
requested merkle trees are stored into a memory cache. (default:
33554432 bytes). The value may be given in bytes or with unit (B,
kiB, MiB, GiB).
.PP
Connection options:
.HP
\fB\-addnode=\fR<ip>
.IP
Add a node to connect to and attempt to keep the connection open
.HP
\fB\-banscore=\fR<n>
.IP
Threshold for disconnecting misbehaving peers (default: 100)
.HP
\fB\-bantime=\fR<n>
.IP
Number of seconds to keep misbehaving peers from reconnecting (default:
86400)
.HP
\fB\-bind=\fR<addr>
.IP
Bind to given address and always listen on it. Use [host]:port notation
for IPv6
.HP
\fB\-blockstallingmindownloadspeed=\fR<n>
.IP
Minimum average download speed (Kbytes/s) we will allow a stalling peer
to fall to during IBD. A value of 0 means stall detection is
disabled (default: 100Kb/s)
.HP
\fB\-broadcastdelay=\fR<n>
.IP
Set inventory broadcast delay duration in millisecond(min: 0, max:
50000)
.HP
\fB\-connect=\fR<ip>
.IP
Connect only to the specified node(s); \fB\-noconnect\fR or \fB\-connect\fR=\fI\,0\/\fR alone to
disable automatic connections
.HP
\fB\-discover\fR
.IP
Discover own IP addresses (default: 1 when listening and no \fB\-externalip\fR
or \fB\-proxy\fR)
.HP
\fB\-dns\fR
.IP
Allow DNS lookups for \fB\-addnode\fR, \fB\-seednode\fR and \fB\-connect\fR (default: 1)
.HP
\fB\-dnsseed\fR
.IP
Query for peer addresses via DNS lookup, if low on addresses (default: 1
unless \fB\-connect\fR/\-noconnect)
.HP
\fB\-externalip=\fR<ip>
.IP
Specify your own public address
.HP
\fB\-forcednsseed\fR
.IP
Always query for peer addresses via DNS lookup (default: 1)
.HP
\fB\-listen\fR
.IP
Accept connections from outside (default: 1 if no \fB\-proxy\fR or
\fB\-connect\fR/\-noconnect)
.HP
\fB\-maxconnections=\fR<n>
.IP
Maintain at most <n> connections to peers (default: 125)
.HP
\fB\-maxreceivebuffer=\fR<n>
.IP
Maximum per\-connection receive buffer in kilobytes (default: 500000).
The value may be given in kilobytes or with unit (B, kB, MB, GB).
.HP
\fB\-maxsendbuffer=\fR<n>
.IP
Maximum per\-connection send buffer in kilobytes (default: 500000). The
value may be given in kilobytes or with unit (B, kB, MB, GB).
.HP
\fB\-maxsendbuffermult=\fR<n>
.IP
Temporary multiplier applied to the \fB\-maxsendbuffer\fR size to allow
connections to unblock themselves in the unlikely situation where
they have become paused for both sending and receiving (default:
10)
.HP
\fB\-factormaxsendqueuesbytes=\fR<n>
.IP
Factor that will be multiplied with excessiveBlockSize to limit the
maximum bytes in all sending queues. If this size is exceeded, no
response to block related P2P messages is sent. (default factor:
4)
.HP
\fB\-maxtimeadjustment\fR
.IP
Maximum allowed median peer time offset adjustment. Local perspective of
time may be influenced by peers forward or backward by this
amount. (default: 4200 seconds)
.HP
\fB\-multistreams\fR
.IP
Enable the use of multiple streams to our peers (default: 1)
.HP
\fB\-multistreampolicies\fR
.IP
List of stream policies to use with our peers in order of preference
(available policies: BlockPriority,Default, default:
BlockPriority,Default)
.HP
\fB\-onlynet=\fR<net>
.IP
Only connect to nodes in network <net> (ipv4 or ipv6)
.HP
\fB\-permitbaremultisig\fR
.IP
Relay non\-P2SH multisig (default: 1)
.HP
\fB\-peerbloomfilters\fR
.IP
Support filtering of blocks and transaction with bloom filters (default:
1)
.HP
\fB\-port=\fR<port>
.IP
Listen for connections on <port> (default: 9883 or testnet: 19883)
.HP
\fB\-proxy=\fR<ip:port>
.IP
Connect through SOCKS5 proxy
.HP
\fB\-proxyrandomize\fR
.IP
Randomize credentials for every proxy connection. (default: 1)
.HP
\fB\-seednode=\fR<ip>
.IP
Connect to a node to retrieve peer addresses, and disconnect
.HP
\fB\-timeout=\fR<n>
.IP
Specify connection timeout in milliseconds (minimum: 1, default: 5000)
.HP
\fB\-upnp\fR
.IP
Use UPnP to map the listening port (default: 0)
.HP
\fB\-whitebind=\fR<addr>
.IP
Bind to given address and whitelist peers connecting to it. Use
[host]:port notation for IPv6
.HP
\fB\-whitelist=\fR<IP address or network>
.IP
Whitelist peers connecting from the given IP address (e.g. 1.2.3.4) or
CIDR notated network (e.g. 1.2.3.0/24). Can be specified multiple
times. Whitelisted peers cannot be DoS banned and their
transactions are always relayed, even if they are already in the
mempool, useful e.g. for a gateway
.HP
\fB\-whitelistrelay\fR
.IP
Accept relayed transactions received from whitelisted peers even when
not relaying transactions (default: 1)
.HP
\fB\-whitelistforcerelay\fR
.IP
Force relay of transactions from whitelisted peers even if they violate
local relay policy (default: 1)
.HP
\fB\-maxuploadtarget=\fR<n>
.IP
Tries to keep outbound traffic under the given target (in MiB per 24h),
0 = no limit (default: 0). The value may be given in megabytes or
with unit (KiB, MiB, GiB).
.PP
Wallet options:
.HP
\fB\-disablewallet\fR
.IP
Do not load the wallet and disable wallet RPC calls
.HP
\fB\-keypool=\fR<n>
.IP
Set key pool size to <n> (default: 1000)
.HP
\fB\-fallbackfee=\fR<amt>
.IP
A fee rate (in SPAVE/kB) that will be used when fee estimation has
insufficient data (default: 0.0002)
.HP
\fB\-mintxfee=\fR<amt>
.IP
Fees (in SPAVE/kB) smaller than this are considered zero fee for
transaction creation (default: 0.00001)
.HP
\fB\-paytxfee=\fR<amt>
.IP
Fee (in SPAVE/kB) to add to transactions you send (default: 0.00)
.HP
\fB\-rescan\fR
.IP
Rescan the block chain for missing wallet transactions on startup
.HP
\fB\-salvagewallet\fR
.IP
Attempt to recover private keys from a corrupt wallet on startup
.HP
\fB\-spendzeroconfchange\fR
.IP
Spend unconfirmed change when sending transactions (default: 1)
.HP
\fB\-usehd\fR
.IP
Use hierarchical deterministic key generation (HD) after BIP32. Only has
effect during wallet creation/first start (default: 1)
.HP
\fB\-upgradewallet\fR
.IP
Upgrade wallet to latest format on startup
.HP
\fB\-wallet=\fR<file>
.IP
Specify wallet file (within data directory) (default: wallet.dat)
.HP
\fB\-walletbroadcast\fR
.IP
Make the wallet broadcast transactions (default: 1)
.HP
\fB\-walletnotify=\fR<cmd>
.IP
Execute command when a wallet transaction changes (%s in cmd is replaced
by TxID)
.HP
\fB\-zapwallettxes=\fR<mode>
.IP
Delete all wallet transactions and only recover those parts of the
blockchain through \fB\-rescan\fR on startup (1 = keep tx meta data e.g.
account owner and payment request information, 2 = drop tx meta
data)
.PP
ZeroMQ notification options:
.HP
\fB\-zmqpubhashblock=\fR<address>
.IP
Enable publish hash block in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubhashtx=\fR<address>
.IP
Enable publish hash transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawblock=\fR<address>
.IP
Enable publish raw block in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawtx=\fR<address>
.IP
Enable publish raw transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubinvalidtx=\fR<address>
.IP
Enable publish invalid transaction in <address>. \fB\-invalidtxsink\fR=\fI\,ZMQ\/\fR
should be specified. For more information see doc/zmq.md.
.HP
\fB\-zmqpubremovedfrommempool=\fR<address>
.IP
Enable publish removal of transaction (txid and the reason in json
format) in <address>. For more information see doc/zmq.md.
.HP
\fB\-zmqpubremovedfrommempoolblock=\fR<address>
.IP
Enable publish removal of transaction (txid and the reason in json
format) in <address>. For more information see doc/zmq.md.
.HP
\fB\-zmqpubhashtx2=\fR<address>
.IP
Enable publish hash transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawtx2=\fR<address>
.IP
Enable publish raw transaction in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubhashblock2=\fR<address>
.IP
Enable publish hash block in <address>. For more information see
doc/zmq.md.
.HP
\fB\-zmqpubrawblock2=\fR<address>
.IP
Enable publish raw block in <address>. For more information see
doc/zmq.md.
.PP
Debugging/Testing options:
.HP
\fB\-uacomment=\fR<cmt>
.IP
Append comment to the user agent string
.HP
\fB\-debug=\fR<category>
.IP
Output debugging information (default: 0, supplying <category> is
optional). If <category> is not supplied or if <category> = 1,
output all debugging information.<category> can be: mempool,
http, zmq, db, rpc, addrman, selectcoins, reindex,
cmpctblock, rand, prune, proxy, mempoolrej, libevent, coindb,
leveldb, txnprop, txnsrc, journal, txnval, netconn, netmsg,
netmsgverb, netmsgall, net, doublespend.
.HP
\fB\-debugexclude=\fR<category>
.IP
Exclude debugging information for a category. Can be used in conjunction
with \fB\-debug\fR=\fI\,1\/\fR to output debug logs for all categories except one
or more specified categories.
.HP
\fB\-help\-debug\fR
.IP
Show all debugging options (usage: \fB\-\-help\fR \fB\-help\-debug\fR)
.HP
\fB\-debugp2pthreadstalls\fR
.IP
Log P2P requests that stall request processing loop for longer than
specified milliseconds (default: disabled)
.HP
\fB\-logips\fR
.IP
Include IP addresses in debug output (default: 0)
.HP
\fB\-logtimestamps\fR
.IP
Prepend debug output with timestamp (default: 1)
.HP
\fB\-minrelaytxfee=\fR<amt>
.IP
Fees (in SPAVE/kB) smaller than this are considered zero fee for relaying,
mining and transaction creation (default: 0.0000025)
.HP
\fB\-maxtxfee=\fR<amt>
.IP
Maximum total fees (in SPACE) to use in a single wallet transaction or raw
transaction; setting this too low may abort large transactions
(default: 0.10)
.HP
\fB\-printtoconsole\fR
.IP
Send trace/debug info to console instead of mvcd.log file
.HP
\fB\-shrinkdebugfile\fR
.IP
Shrink mvcd.log file on client startup (default: 1 when no \fB\-debug\fR)
.PP
Chain selection options:
.HP
\fB\-testnet\fR
.IP
Use the test chain
.HP
\fB\-regtest\fR
.IP
Enter regression test mode, which uses a special chain in which blocks
can be solved instantly. This is intended for regression testing
tools and app development.
.HP
\fB\-stn\fR
.IP
Use the Scaling Test Network
.PP
Node relay options:
.HP
\fB\-datacarrier\fR
.IP
Relay and mine data carrier transactions (default: 1)
.HP
\fB\-datacarriersize\fR
.IP
Maximum size of data in data carrier transactions we relay and mine
(default: 4294967295). The value may be given in bytes or with
unit (B, kB, MB, GB).
.HP
\fB\-maxstackmemoryusagepolicy\fR
.IP
Set maximum stack memory usage used for script verification we're
willing to relay/mine in a single transaction (default: 100 MB, 0
= unlimited) after Genesis is activated (policy level). The value
may be given in bytes or with unit (B, kB, MB, GB). \fR.
.HP
\fB\-maxopsperscriptpolicy=\fR<n>
.IP
Set maximum number of non\-push operations we're willing to relay/mine
per script (default: unlimited, 0 = unlimited), after Genesis is
activated
.HP
\fB\-maxtxsigopscountspolicy=\fR<n>
.IP
Set maximum allowed number of signature operations we're willing to
relay/mine in a single transaction (default: unlimited, 0 =
unlimited) after Genesis is activated.
.HP
\fB\-maxstdtxvalidationduration=\fR<n>
.IP
Set the single standard transaction validation duration threshold in
milliseconds after which the standard transaction validation will
terminate with error and the transaction is not accepted to
mempool (min 1ms, default: 3ms)
.HP
\fB\-maxnonstdtxvalidationduration=\fR<n>
.IP
Set the single non\-standard transaction validation duration threshold in
milliseconds after which the non\-standard transaction validation
will terminate with error and the transaction is not accepted to
mempool (min 10ms, default: 1000ms)
.HP
\fB\-maxtxchainvalidationbudget=\fR<n>
.IP
Set the upper limit of unused validation time to add to the next
transaction validated in the chain (min 0ms, default: 50ms)
.HP
\fB\-validationclockcpu\fR
.IP
Use CPU time instead of wall clock time for validation duration
measurement (default: 1)
.HP
\fB\-maxtxsizepolicy=\fR<n>
.IP
Set maximum transaction size in bytes we relay and mine (default: 10 MB,
min: 99999 B, 0 = unlimited) after Genesis is activated. The
value may be given in bytes or with unit (B, kB, MB, GB).
.HP
\fB\-minconsolidationfactor=\fR<n>
.IP
Set minimum ratio between sum of utxo scriptPubKey sizes spent in a

doc/release-notes.md
# MVC version 0.1.4 Release Notes

Added ASERT difficulty adjustment algorithm with two days half-life, aka, aserti3-2d. 
This upgrade uses a MTP-based activation,1686636000(2023-06-13 06:00:00 +0000) for mainnet and 1685426400(2023-05-30 06:00:00 +0000) for testnet.
The block before the activation block will be used as the anchor block for the ASERT algorithm.
It is highly recommended to upgrade this update, otherwise it will fork after ASERT activation.

Added parameter minrelaytxfee. 
Fees smaller than which are considered zero fee for relaying, mining and transaction creation (default: 250). 
If you set it to zero then a transaction spammer can cheaply fill blocks using 1-satoshi-fee transactions, so which is not allowed. 
If you are mining, be careful setting this, it should be set above the real cost to you of processing a transaction.

Upgraded minimum chain work and added new checkpoints;

Bug fixed and redundant code removed;

Update version to v0.1.4.0



# MVC version 0.1.3.1 Release Notes

Remove dustlimitfactor and dustrelayfee settings.

Hard code dust threshold to zero to allow 1 sat transaction outputs.

Bug fix.




# MVC Node software – Upgrade to v0.1.3 Release

Significant improvement in the performance of node validation of transactions.

Improved node performance for processing smart contracts.

New RPC method for receiving UTXOs.

Increased the default limit for node processing of smart contract scripts.

Significant improvement in the ability of nodes to parallel process transactions and smart contracts.

Allows parallel shard downloads of the same block from multiple nodes.

Enhances the connectivity of the MVC network.

Notification information for new block double-spending.

Added Webhook interface for double-spending and fork detection.

Removed the ancestor chain quantity limit.

Added RPC to provide Merkle proof that transactions are included in blocks.

Fixed known issues in many past versions.

src/mvcd.cpp
// Copyright (c) 2009-2010 Satoshi Nakamoto
// Copyright (c) 2009-2021 The Bitcoin Core developers
// Copyright (c) 2021-2023 The MVC developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#if defined(HAVE_CONFIG_H)
#include "config/mvc-config.h"
#endif

#include "chainparams.h"
#include "clientversion.h"
#include "compat.h"
#include "config.h"
#include "fs.h"
#include "init.h"
#include "noui.h"
#include "scheduler.h"
#include "util.h"
#include "utilstrencodings.h"
#include <boost/algorithm/string/predicate.hpp>
#include <boost/thread.hpp>

#include <cstdio>


/* Introduction text for doxygen: */

/*! \mainpage Developer documentation
 *
 * \section intro_sec Introduction
 *
 * This is the developer documentation of MicroVisionChain
 * (https://www.microvisionchain.com/). MVC is a client for the digital
 * currency called MicroVisionChain, which enables
 * instant payments to anyone, anywhere in the world. MicroVisionChain uses
 * peer-to-peer technology to operate with no central authority: managing
 * transactions and issuing money are carried out collectively by the network.
 *
 * \section Navigation
 * Use the buttons <code>Namespaces</code>, <code>Classes</code> or
 * <code>Files</code> at the top of the page to start navigating the code.
 */

void WaitForShutdown(boost::thread_group *threadGroup, const task::CCancellationToken& shutdownToken) {

    // Tell the main threads to shutdown.
    while (!shutdownToken.IsCanceled()) {
        MilliSleep(200);
    }
    if (threadGroup) {
        Interrupt(*threadGroup);
        threadGroup->join_all();
    }
}

//////////////////////////////////////////////////////////////////////////////
//
// Start
//
bool AppInit(int argc, char *argv[]) {
    RenameThread("main");
    boost::thread_group threadGroup;
    CScheduler scheduler;

    // FIXME: Ideally, we'd like to build the config here, but that's currently
    // not possible as the whole application has too many global state. However,
    // this is a first step.
    auto& config = GlobalConfig::GetModifiableGlobalConfig();

    bool fRet = false;

    //
    // Parameters
    //
    // main()
    gArgs.ParseParameters(argc, argv);

    // Process help and version before taking care about datadir
    if (gArgs.IsArgSet("-?") || gArgs.IsArgSet("-h") ||
        gArgs.IsArgSet("-help") || gArgs.IsArgSet("-version")) {
        std::string strUsage = strprintf(_("%s"), _(PACKAGE_NAME)) +
                               " " + _("version") + " " + FormatFullVersion() +
                               "\n";

        if (gArgs.IsArgSet("-version")) {
            strUsage += FormatParagraph(LicenseInfo());
        } else {
            strUsage += "\n" + _("Usage:") + "\n" +
                        "  mvcd [options]                     " +
                        strprintf(_("Start %s Daemon"), _(PACKAGE_NAME)) + "\n";

            strUsage += "\n" + HelpMessage(HMM_MVCD, config);
        }

        fprintf(stdout, "%s", strUsage.c_str());
        return true;
    }

    try {
        if (!fs::is_directory(GetDataDir(false))) {
            fprintf(stderr,
                    "Error: Specified data directory \"%s\" does not exist.\n",
                    gArgs.GetArg("-datadir", "").c_str());
            return false;
        }
        try {
            gArgs.ReadConfigFile(gArgs.GetArg("-conf", MVC_CONF_FILENAME));
        } catch (const std::exception &e) {
            fprintf(stderr, "Error reading configuration file: %s\n", e.what());
            return false;
        }
        // Check for -testnet or -regtest parameter (Params() calls are only
        // valid after this clause)
        try {
            SelectParams(ChainNameFromCommandLine());
        } catch (const std::exception &e) {
            fprintf(stderr, "Error: %s\n", e.what());
            return false;
        }

        // Fill config with block size data
        config.SetDefaultBlockSizeParams(Params().GetDefaultBlockSizeParams());

        // Command-line RPC
        bool fCommandLine = false;
        for (int i = 1; i < argc; i++)
            if (!IsSwitchChar(argv[i][0]) &&
                !boost::algorithm::istarts_with(argv[i], "mvc:"))
                fCommandLine = true;

        if (fCommandLine) {
            fprintf(stderr, "Error: There is no RPC client functionality in "
                            "mvcd anymore. Use the mvc-cli utility "
                            "instead.\n");
            exit(EXIT_FAILURE);
        }
        // -server defaults to true for mvcd
        gArgs.SoftSetBoolArg("-server", true);
        // Set this early so that parameter interactions go to console
        InitLogging();
        InitParameterInteraction();
        if (!AppInitBasicSetup()) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (!AppInitParameterInteraction(config)) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (!AppInitSanityChecks()) {
            // InitError will have been called with detailed error, which ends
            // up on console
            exit(1);
        }
        if (gArgs.GetBoolArg("-daemon", false)) {
#if HAVE_DECL_DAEMON

            fprintf(stdout, "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m╭━╮╭━╮╭╮  ╭╮╭━━━╮\033[0m\033[1;40;34m▒▒▒▒\033[0m\033[1;40;31mMicroVision!\033[0m\033[1;40;34m▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃┃╰╯┃┃┃╰╮╭╯┃┃╭━╮┃\033[0m\033[1;40;34m▒▒▒▒  \033[0m\033[1;40;31mv0.1.4.0\033[0m\033[1;40;34m  ▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃╭╮╭╮┃╰╮┃┃╭╯┃┃ ╰╯\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃┃┃┃┃┃ ┃╰╯┃ ┃┃ ╭╮\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃┃┃┃┃┃ ╰╮╭╯ ┃╰━╯┃\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m╰╯╰╯╰╯  ╰╯  ╰━━━╯\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m╭━━━╮╭━━━╮╭━━━╮╭╮  ╭╮╭━━━╮╭━━━╮\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃╭━╮┃┃╭━━╯┃╭━╮┃┃╰╮╭╯┃┃╭━━╯┃╭━╮┃\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃╰━━╮┃╰━━╮┃╰━╯┃╰╮┃┃╭╯┃╰━━╮┃╰━╯┃\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m╰━━╮┃┃╭━━╯┃╭╮╭╯ ┃╰╯┃ ┃╭━━╯┃╭╮╭╯\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃╰━╯┃┃╰━━╮┃┃┃╰╮ ╰╮╭╯ ┃╰━━╮┃┃┃╰╮\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\033[1;40;32m╰━━━╯╰━━━╯╰╯╰━╯  ╰╯  ╰━━━╯╰╯╰━╯\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒\033[0m\033[1;40;32m╭━━━╮╭━━━━╮╭━━━╮╭━━━╮╭━━━━╮╭━━╮╭━╮ ╭╮╭━━━╮\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃╭━╮┃┃╭╮╭╮┃┃╭━╮┃┃╭━╮┃┃╭╮╭╮┃╰┫┣╯┃┃╰╮┃┃┃╭━╮┃\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃╰━━╮╰╯┃┃╰╯┃┃ ┃┃┃╰━╯┃╰╯┃┃╰╯ ┃┃ ┃╭╮╰╯┃┃┃ ╰╯\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒\033[0m\033[1;40;32m╰━━╮┃  ┃┃  ┃╰━╯┃┃╭╮╭╯  ┃┃   ┃┃ ┃┃╰╮┃┃┃┃╭━╮\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒\033[0m\033[1;40;32m┃╰━╯┃  ┃┃  ┃╭━╮┃┃┃┃╰╮  ┃┃  ╭┫┣╮┃┃ ┃┃┃┃╰┻━┃\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒\033[0m\033[1;40;32m╰━━━╯  ╰╯  ╰╯ ╰╯╰╯╰━╯  ╰╯  ╰━━╯╰╯ ╰━╯╰━━━╯\033[0m\033[1;40;34m▒▒▒▒▒▒▒▒\033[0m\n"
                            "\033[1;40;34m▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒▒\033[0m\n");

            // Daemonize
            if (daemon(1, 0)) {
                // don't chdir (1), do close FDs (0)
                fprintf(stderr, "Error: daemon() failed: %s\n",
                        strerror(errno));
                return false;
            }
#else
            fprintf(
                stderr,
                "Error: -daemon is not supported on this operating system\n");
            return false;
#endif // HAVE_DECL_DAEMON
        }

        fRet = AppInitMain(config, threadGroup, scheduler, GetShutdownToken());
    } catch (const std::exception &e) {
        PrintExceptionContinue(&e, "AppInit()");
    } catch (...) {
        PrintExceptionContinue(nullptr, "AppInit()");
    }

    if (!fRet) {
        Interrupt(threadGroup);
        // threadGroup.join_all(); was left out intentionally here, because we
        // didn't re-test all of the startup-failure cases to make sure they
        // don't result in a hang due to some
        // thread-blocking-waiting-for-another-thread-during-startup case.
    } else {
        LogPrintf("Preload wait for shutdown\n");
        WaitForShutdown(&threadGroup, GetShutdownToken());
        LogPrintf("Preload wait for shutdown done\n");
    }
    LogPrintf("Checking Thread shutdown\n");
    Shutdown();

    return fRet;
}

int main(int argc, char *argv[]) {
    SetupEnvironment();

    // Connect mvcd signal handlers
    noui_connect();

    return (AppInit(argc, argv) ? EXIT_SUCCESS : EXIT_FAILURE);
}
